src/execution_plan/ops/op_node_hash_join.c

```diff
@@ -0,0 +1,244 @@
+ * node stored at postion intersect_node_idx. */
+static bool _record_islt(Record l, Record r, uint intersect_node_idx) {
+    Node *ln = Record_GetNode(l, intersect_node_idx);
+    Node *rn = Record_GetNode(r, intersect_node_idx);
+    return ENTITY_GET_ID(ln) < ENTITY_GET_ID(rn);
+}
+
+/* `intersect_node_idx` is an actual variable in the caller function.
+ * Using it in a macro like this is rather ugly,
+ * but the macro passed to QSORT must accept only 2 arguments. */
+#define RECORD_SORT_ON_ENTRY(a, b) (_record_islt((*a), (*b), intersect_node_idx))
+
+/* Retrive the next intersecting record
+ * if such exists, otherwise returns NULL. */
+static Record _get_intersecting_record(OpNodeHashJoin *op) {
+    // No more intersecting records.
+    if(op->intersect_idx == NO_INTERSECTION ||
+       op->number_of_intersections == 0) return NULL;
+
+    Record cr = op->cached_records[op->intersect_idx];
+    
+    // Update intersection trackers.
+    op->intersect_idx++;
+    op->number_of_intersections--;
+
+    return cr;
+}
+
+static uint64_t _binarySearch(const OpNodeHashJoin *op, NodeID id) {
+    Record *array = op->cached_records;
+    uint32_t recordCount = array_len(array);
+    uint32_t pos;
+    uint32_t left = 0;
+    uint32_t right = recordCount;
+    while(left < right) {
+        pos = (right + left) / 2;
+        Node *n = Record_GetNode(array[pos], op->intersect_node_idx);
+        if(ENTITY_GET_ID(n) < id) left = pos + 1;
+        else right = pos;
+    }
+    return left;
+}
+
+/* Look up first cached record CR position
+ * where CR node ID = id.
+ * Returns NO_INTERSECTION if no such CR is found. */
+int _set_intersection_idx(OpNodeHashJoin *op, NodeID id) {
+    op->intersect_idx = NO_INTERSECTION;
+    op->number_of_intersections = 0;
+    int idx = _binarySearch(op, id);
+
+    // Make sure node was found.
+    Record r = op->cached_records[idx];
+    Node *n = Record_GetNode(r, op->intersect_node_idx);
+
+    // Node was not found.
+    if(ENTITY_GET_ID(n) != id) return NO_INTERSECTION;
+
+    /* Node was found, make sure op->intersect_idx
+     * points to the first record holding node, 
+     * and update number_of_intersections to count how many
+     * records share the same node ID. */
+    while(idx >= 0) {
+        r = op->cached_records[idx];
+        Node *n = Record_GetNode(r, op->intersect_node_idx);
+        if(ENTITY_GET_ID(n) != id) break;
+        idx--;
+    }
+
+    op->intersect_idx = idx;
+    if(idx > 0) op->intersect_idx++;
+    idx = op->intersect_idx;
+
+    r = op->cached_records[idx];
+    n = Record_GetNode(r, op->intersect_node_idx);
+    uint record_count = array_len(op->cached_records);
+
+    // Count how many records share the same node.
+    while(idx < record_count) {
+        r = op->cached_records[idx];
+        n = Record_GetNode(r, op->intersect_node_idx);
+        if(ENTITY_GET_ID(n) != id) break;
+        op->number_of_intersections++;
+    }
```

Comment:
Instead of this loop, why don't we run another binary search using `idx` as one of the bounds?

I think it would be cleanest to have the search at 64 find the leftmost index, then use that as the left bound here in a find-rightmost binary search.